Encryption/signature confusion (#8)
yaronf/draft-sheffer-oauth-rfc8725bis: draft-sheffer-oauth-rfc8725bis.md

## Incorrect Use and Composition of Encryption and Signature {#incorrect-composition-of-encryption-and-signature}

> Most authentication use cases only require a simple signed JWT as their token. However verifiers don't always check that the received JWT is a signed JWS as opposed to an encrypted JWE structure. This can result in vulnerabilities, in particular when the verifier's asymmetric key includes the private key, and when the verifier's library does not distinguish between successful decryption and successful signature validation {{CVE-2023-51774}}.

Reviewer: I don't understand what the phrase "when the verifier's asymmetric key includes the private key" is getting at.  How does this relate to the vulnerability?

Author: On the receiver side, if all you want is a JWS then you only need a public key in the JWK, to verify the signature.

But if you're accepting a JWE, you also need a private key in the JWK to decrypt the thing. So the attack (sending a JWE instead of a JWS) only works if there's a private key on the receiver's side.

> Most authentication use cases only require a simple signed JWT as their token. However verifiers don't always check that the received JWT is a signed JWS as opposed to an encrypted JWE structure. This can result in vulnerabilities, in particular when the verifier's asymmetric key includes the private key, and when the verifier's library does not distinguish between successful decryption and successful signature validation {{CVE-2023-51774}}.

Reviewer: ```suggestion
Most authentication use cases only require a simple signed JWT as their token. However verifiers don't always check that the received JWT is a signed JWS as opposed to an encrypted JWE structure. This can result in vulnerabilities when the verifier's library does not distinguish between successful decryption and successful signature validation {{CVE-2023-51774}}.
```

Delete the ambiguous phrase

Reviewer: Please apply my suggestion deleting the ambiguous phrase, then I will approve.

Reviewer 2: I'm also fine with Mike's phrasing ...

## Validate All Cryptographic Operations {#validate-crypto}

All cryptographic operations used in the JWT MUST be validated and the entire JWT MUST be rejected
if any of them fail to validate.
This is true not only of JWTs with a single set of Header Parameters
but also for Nested JWTs in which both outer and inner operations  MUST be validated
using the keys and algorithms supplied by the application.

> Libraries MUST allow the verifier to distingush between JWS-signed and JWE-encrypted JWTs, and to only accept the former.

Reviewer: ```suggestion
Libraries MUST allow the verifier to distinguish between JWS-signed and JWE-encrypted JWTs, enabling callers to accept only JWTs that meet their security criteria.
```

Author: That would allow the verifier to only accept JWS, if that's their policy.

> When some claims of the JWT require confidentiality protection, it is RECOMMENDED to nest encrypted JWEs within signed JWSs instead of vice versa.

Reviewer: ```suggestion
```

This is unrelated to the other changes in this PR.  Please remove it.

Author: For the record, the following in RFC 7519 convinced me:

> Note that potential concerns about security issues related to the order of signing and encryption operations are already addressed by the underlying JWS and JWE specifications; in particular, because JWE **only supports the use of authenticated encryption algorithms**, cryptographic concerns about the potential need to sign after encryption that apply in many contexts do not apply to this specification.

> When some claims of the JWT require confidentiality protection, it is RECOMMENDED to nest encrypted JWEs within signed JWSs instead of vice versa.

Author: Move into a separate PR.
